whittle/lora.py

```diff
@@ -0,0 +1,821 @@
+# Copyright Lightning AI. Licensed under the Apache License 2.0, see LICENSE file.
+from __future__ import annotations
+
+import dataclasses
+import math
+import os
+import time
+from datetime import datetime
+from pathlib import Path
+from pprint import pprint
+from typing import Any, Literal
+
+import lightning as L
+import torch
+import yaml  # type: ignore
+from lightning.fabric.plugins import BitsandbytesPrecision
+from lightning.fabric.strategies import DDPStrategy
+from litgpt.args import EvalArgs, TrainArgs
+from litgpt.data import Alpaca, DataModule
+from litgpt.generate.base import generate
+from litgpt.lora import (
+    lora_filter,
+    mark_only_lora_as_trainable,
+)
+from litgpt.prompts import save_prompt_style
+from litgpt.tokenizer import Tokenizer
+from litgpt.utils import (
+    CycleIterator,
+    auto_download_checkpoint,
+    check_nvlink_connectivity,
+    check_valid_checkpoint_dir,
+    choose_logger,
+    chunked_cross_entropy,
+    copy_config_files,
+    get_default_supported_precision,
+    init_out_dir,
+    instantiate_bnb_optimizer,
+    instantiate_torch_optimizer,
+    load_checkpoint,
+    num_parameters,
+    parse_devices,
+    save_hyperparameters,
+)
+from torch.utils.data import ConcatDataset, DataLoader
+from torchmetrics import RunningMean
+
+from search.search_spaces import search_spaces
+from whittle.args import SamplerArgs
+from whittle.data.llamamini import LLaMaMini
+from whittle.eval.utils import compute_accuracy
+from whittle.lora_model.config import LoRAConfig as Config
+from whittle.lora_model.lora_gpt import GPT
+from whittle.lora_model.merge import merge_lora
+from whittle.sampling.samplers import get_sampler
+from whittle.training_strategies.base_strategy import BaseTrainingStrategy
+from whittle.training_strategies.sandwich import SandwichStrategy
+from whittle.training_strategies.standard import StandardStrategy
+
+# Copyright Lightning AI. Licensed under the Apache License 2.0, see LICENSE file.
+
+"""This script merges the LoRA weights with the base model"""
+
+
+def compute_loss(model, val_dataloader, eval):
+    losses = torch.zeros(min(len(val_dataloader), eval.max_iters))
+    for k, batch in enumerate(val_dataloader):
+        if k >= eval.max_iters:
+            break
+        input_ids, targets = batch["input_ids"], batch["labels"]
+        logits = model(input_ids)
+        losses[k] = chunked_cross_entropy(
+            logits[..., :-1, :], targets[..., 1:], chunk_size=0
+        )
+
+    val_loss = losses.mean()
+    return val_loss
+
+
+def plot_validation_metrics(model, val_dataloader, eval, sampler):
```

Comment:
Could we rename this function (e.g compute validation_metrics_subnets or similar) since we only provide the data for plotting but do not plot directly